Get started guide

eu-siemann opened the pull request.

merltron committed:
adding base template files

adding converted "developer guide" template files for developers to update

Commit:
Adding full asciidoc content templates

Adding converted asciidoc files with commentary
inside them. Also updated config to integrate into
Jekyll navigation for local testing.

Commit:
Making template topics more OTA Specific

Added examples of code snippets and doxygen links
Consolidated everything into one section "Using libaktualizr"
Added more relevant "inspiration" links.

Commit:
Using libaktualizr: add get started guide

Signed-off-by: Eugene Smirnov <evgenii.smirnov@here.com>

xaviergully commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+under the Mozilla Public License 2.0 and is located on github:
+https://github.com/advancedtelematic/aktualizr
+
+The project is intended to be build on a Linux system and compiler must support
```
built

xaviergully commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+under the Mozilla Public License 2.0 and is located on github:
+https://github.com/advancedtelematic/aktualizr
+
+The project is intended to be build on a Linux system and compiler must support
```
the project is intented to be built ....

pattivacek replied:
"and _the_ compiler must support"

xaviergully commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+is in executed state.
+
+Can also be called on a previously paused instance, but doesn't change
+it's paused state.
```
its paused state

MatthiasBerchner replied:
If I call abort, the Download status including pause should be reset. To call afterward resume to reset the pause status is a little bit strange, it will not really resume the download. Without the resume additionally new Downloads are still paused.

xaviergully reviewed (commented):
Small typos, otherwise looks good to me

lbonn commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+$ cd aktualizr
+$ mkdir build
+$ cd build
+$ cmake ..
```
I would still advise using `-DCMAKE_BUILD_TYPE=Release` explicitly, not to rely too much on current defaults.

eu-siemann replied:
makes sense

lbonn commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+and currently running OTA campaigns, downloading and installing the updates,
+reporting the installation results, and so on.
+
+The main library header is `src/libaktualizr/primary/aktualizr.h` and it also
```
It would be cool if paths to project files could link to their current version (master or some stable release). The guide will then be much easier to follow.

lbonn commented:
I think we avoid manually line wrapping our adoc documents, because it usually clutters diffs unnecessarily.

@tkfu or @merltron could maybe confirm.

lbonn commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+[source,cpp]
+Aktualizr::~Aktualizr(Config &config)
+
+Calls Abort() and destroys the aktualizr object.
```
It's a good idea to point out that `Abort()` is called in the destructor but, presented like that, it looks like it's a method of the api that can be used manually. Seasoned C++ developers won't be fooled but maybe have a note to detail that it shouldn't be called manually in most usages.

eu-siemann replied:
I think I'll move this text to the Abort() call description.

lbonn commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+Install previously downloaded targets.
+
+[source,cpp]
+std::ifstream Aktualizr::GetStoredTarget(const Uptane::Target &target)
```
No description: is it intended or waiting for the api to be more clearly defined?

eu-siemann replied:
Oops, lost it somewhere.

lbonn commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+[source,cpp]
+std::future<void> Aktualizr::RunForever()
+
+Asynchronously run aktualizr indefinitely until Shutdown is called.
```
I noticed this is the line from the doxygen but it's maybe too vague. Something like: "Automatic check and install updates indefinitely: runs UptaneCycle() in a loop at regular intervals until Shutdown() is called" ?

eu-siemann replied:
Yes, that sounds better. Except that here we don't have the Shutdown call :)

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+:icons: font
+
+Aktualizr is a {cpp} application that implements client-side functionality for
+HERE OTA Connect system based on the Uptane security framework<<anchor-1,[1]>>.
```
"client-side functionality for _the_ HERE OTA Connect system"

merltron replied:
I find this introduction slightly confusing. I would like to more clearly delineate the client (aktualizr) from the library (libaktualizr). Is this guide describing how to use both? I thought it was specifically for "libaktualizr". I can propose a rewrite in a bit...

pattivacek replied:
This is intended to be about libaktualizr. aktualizr (or really aktualizr-primary) is just an example of an API user.

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+Aktualizr is a {cpp} application that implements client-side functionality for
+HERE OTA Connect system based on the Uptane security framework<<anchor-1,[1]>>.
+While it is possible to use Aktualizr as a standalone CLI application,
+using it as a {cpp} library from a custom application provides user with
```
"provides _the_ user"

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+more flexibility and customization.
+
+This guide focuses on integrating Aktualizr functionality into a
+custom application, in other words it focuses on using libaktualizr.
```
I'd simplify a bit and say something like "...into a custom application by using the libaktualizr API."

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+
+== How to build libaktualizr
+
+Aktualizr is an open-source software distributed
```
"an open-source software _project_ distributed"

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+== Integrating libaktualizr into your application
+
+=== Building libaktualizr in-tree
+The easiest way to integrate with libaktualizr, is to add it as a subdirectory
```
Comma is unnecessary.

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+`aktualizr_static_lib` target. When building your application these libraries
+should be added with
+[source,cmake]
+target_link_libraries(your-app aktualizr_static_lib ${AKTUALIZR_EXTERNAL_LIBS})
```
"This will automatically add _the_ needed entries to `INCLUDE_DIRECTORIES` variable, create _the_ `AKTUALIZR_EXTERNAL_LIBS` variable, which contains the list of external libraries required for linking with libaktualizr, and ~~will~~ create _the_ `aktualizr_static_lib` target."

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+
+Libaktualizr provides a C++ API for fetching information about available updates
+and currently running OTA campaigns, downloading and installing the updates,
+reporting the installation results, and so on.
```
I'd drop the "and so on" and just end with "_and_ reporting the installation results."

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+and add `#include "primary/aktualizr.h"` to your source file.
+
+Most of the API calls, unless specified otherwise, are asynchronous and return
+`std::future` that will contain corresponding result type. Result types are
```
" and return _an_ `std::future`"

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+in sequential order in a separate thread. If the execution was paused, newly
+issued commands will accumulate in the command queue and it's up to the caller
+to ensure that the queue does not get overloaded with unnecessary command
+duplicates.
```
I prefer "unnecessary duplicate commands."

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+duplicates.
+
+In case of failure to perform an action or some other kind of an error an
+exception will be thrown. For asynchronous calls the exception will be stored
```
I think a comma after "kind of an error" would be helpful. And actually the "an" before "error" is extraneous.

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+In case of failure to perform an action or some other kind of an error an
+exception will be thrown. For asynchronous calls the exception will be stored
+in the shared state associated with std::future. All exceptions are subclassed
+from AktualizrException class.
```
"from _the_ AktualizrException class"

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+from AktualizrException class.
+
+There is a sample application that shows how libaktualizr API can be used.
+The application is located in `src/hmi_stub`.
```
Is it worth mentioning that aktualizr-primary itself is an example? It's actually somewhat better than hmi-stub in some regards.

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+All commands that were scheduled after the currently executed command will wait
+in the command queue until Resume() is issued.
+Commands that are issued after Pause() will be put on a command queue,
+but not executed until Resume() is called.
```
Realistically only downloads can be paused, right? All other commands will finish running if they are in progress when the Pause() is issued. The phrasing is fine as it is (I think), but I want to be sure we're on the same page about what should happen.

eu-siemann replied:
Yes, that is exactly what I was trying to say.

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+The Pause() function returns immediately, while pausing the running command
+still may be in progress.
+
+The function has no effect in case execution was already paused.
```
"if" is perhaps better than "in case".

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+
+Requests the currently running command to abort and flushes the command queue.
+The Abort() function will block until the command queue is empty and no command
+is in executed state.
```
"executed state" sounds odd to me. I'd prefer "...until the command queue is empty and all currently executing commands have finished." Or something like that. Is "stopped" better than "finished"?

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+
+Accept a campaign for the current device.
+Campaigns are a concept outside of Uptane, and allow for user approval of
+updates before the contents of the update are known.
```
The second sentence seems redundant since we just said it for CampaignCheck.

pattivacek commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+std::future<void> Aktualizr::SendDeviceData()
+
+Sends local device data to the server.
+This includes network status, installed packages, hardware etc.
```
I'd say "network status, installed packages, _and_ hardware _information_." That's all it is, so the "etc" is extraneous.

pattivacek reviewed (commented):
A lot of nitpicking, sorry. Since this is going to a customer in short order, I'm being a bit more particular than I might otherwise be.

merltron commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+
+The project is intended to be build on a Linux system and compiler must support
+C++11 version of the standard.
+File README.adoc located in the project root directory contains
```
We should should just link to the actual file. Will propose a change..

merltron commented on _posts/2019-01-30-get-started-libaktualizr.adoc:
```diff
@@ -0,0 +1,264 @@
+
+== Integrating libaktualizr into your application
+
+=== Building libaktualizr in-tree
```
"in-tree" and "out-of-tree" seems to be technical  jargon, which is fine. But will non-native speakers and developers who don't use the same tools as you understand what you mean? Just need to double check.

merltron committed:
OTA-1727: Adding revision for getting started

Add a bit of "what is" and "why" info
Tidied up the API reference and tried to put it
in a table that's more scannable.
Fixed some code formatting.

merltron committed:
OTA-1727: Adding revision for getting started
Add a bit of "what is" and "why" info
Tidied up the API reference and tried to put it
in a table that's more scannable.

merltron committed:
Merge branch 'feat/OTA-1727/get-started' of https://github.com/advancedtelematic/ats-garage-docs into feat/OTA-1727/get-started

merltron committed:
Merge branch 'feat/OTA-1727/get-started' of https://github.com/advancedtelematic/ats-garage-docs into feat/OTA-1727/get-started

merltron committed:
Merge branch 'feat/OTA-1727/get-started' of https://github.com/advancedtelematic/ats-garage-docs into feat/OTA-1727/get-started

Commit:
Using libaktualizr: some fixes based on the feedback

Signed-off-by: Eugene Smirnov <evgenii.smirnov@here.com>

Commit:
Using libaktualizr: remove destructor description

Signed-off-by: Eugene Smirnov <evgenii.smirnov@here.com>

Commit:
Using libaktualizr: typo fix

Signed-off-by: Eugene Smirnov <evgenii.smirnov@here.com>

Commit:
Using libaktualizr: add more links

Signed-off-by: Eugene Smirnov <evgenii.smirnov@here.com>

tkfu committed:
Fix symlink to unfuck git merge

Signed-off-by: Jon Oster <jon.oster@here.com>

tkfu committed:
Remove hard-wrapped line endings

Signed-off-by: Jon Oster <jon.oster@here.com>

tkfu commented:
Closing, superseded by #79.
